src/psyclone/parse/file_info.py

```diff
@@ -77,27 +143,286 @@ def filename(self):
-    @property
-    def contents(self):
-        '''Returns the contents of the file. The first time, it
+    def get_source_code(self, verbose: bool = False) -> str:
+        '''Returns the source code of the file. The first time, it
         will be read from the file, but the data is then cached.
 
         If any decoding errors are encountered then the associated character(s)
         are simply skipped. This is because this class is intended for reading
         Fortran source and the only way such characters can appear is if they
         are in comments.
 
+        :param verbose: If `True`, produce some verbose output
+
         :returns: the contents of the file (utf-8 encoding).
-        :rtype: str
 
         '''
         if self._source_code is None:
             # Specifying errors='ignore' simply skips any characters that
             # result in decoding errors. (Comments in a code may contain all
             # sorts of weird things.)
-            with open(self._filename, "r", encoding='utf-8',
-                      errors='ignore') as file_in:
-                self._source_code = file_in.read()
+
+            if verbose:
+                print(
+                    f"- Source file '{self._filename}': "
+                    f"Loading source code"
+                )
+
+            try:
+                with open(
+                    self._filename, "r", encoding="utf-8", errors="ignore"
+                ) as file_in:
+                    self._source_code = file_in.read()
+            except FileNotFoundError:
+                raise FileNotFoundError(
+                    f"No such file or directory: '{self._filename}'"
+                )
+
+            # Compute hash sum which will be used to check cache of fparser
+            self._source_code_hash_sum = hashlib.md5(
+                self._source_code.encode()
+            ).hexdigest()
 
         return self._source_code
+
+    @property
+    def source_code(self, verbose: bool = False) -> str:
+        return self.get_source_code(verbose)
+
+    def _cache_load(
+        self,
+        verbose: bool = False,
+    ):
+        """Load data from the cache file if possible.
+        This also checks for matching checksums after loading the data
+        from the cache.
+
+        :param verbose: Produce some verbose output
+        """
+
+        if not self._use_caching:
+            return
+
+        # Load the source code in case it's not yet loaded.
+        # This also fills in the hash sum
+        self.get_source_code()
+
+        assert self._source_code_hash_sum is not None, (
+            "Source code needs to be loaded before fparser or psyir "
+            "representation is loaded."
+        )
+
+        # Check whether cache was already loaded
+        if self._cache_data_load is not None:
+            return self._cache_data_load
+
+        # Load cache file
+        try:
+            filehandler = open(self._filepath_cache, "rb")
```

Comment:
This bit too will need some thought for parallel builds. Please add a comment.

Reply:
OK.